libibverbs/dummy_ops.c

```diff
@@ -369,11 +369,22 @@ static int query_device(struct ibv_context *context,
 	return EOPNOTSUPP;
 }
 
+/* Provide a generic implementation for all providers that don't implement
+ * query_device_ex.
+ */
 static int query_device_ex(struct ibv_context *context,
 			   const struct ibv_query_device_ex_input *input,
 			   struct ibv_device_attr_ex *attr, size_t attr_size)
 {
-	return EOPNOTSUPP;
+	if (input && input->comp_mask)
```

Comment:
don't test for null, crashing is the right thing to do

Reply:
Actually, I'm not sure about the comp_mask any more, why doesn't the wrapper test it?
This is one of those things that doesn't make any sense :(
Stick with copying the wrapper, no sense in making it more complicated if we ever want to use this thing.

Reply:
Why is crashing the right thing to do? 

which wrapper are you referring to? currently the ibv_cmd_query_device_ex is checking comp_mask -> do you want to take it out and place it in the ibv_query_device_ex?

Reply:
Sigh, this is a big mess it seems.

There is already a fallback wrapper for query_device_ex in verbs.h:

```
static inline int
ibv_query_device_ex(struct ibv_context *context,
                    const struct ibv_query_device_ex_input *input,
                    struct ibv_device_attr_ex *attr)
{
        struct verbs_context *vctx;
        int ret;

        vctx = verbs_get_ctx_op(context, query_device_ex);
        if (!vctx)
                goto legacy;

        ret = vctx->query_device_ex(context, input, attr, sizeof(*attr));
        if (ret == EOPNOTSUPP)
                goto legacy;

        return ret;

legacy:
        memset(attr, 0, sizeof(*attr));
        ret = ibv_query_device(context, &attr->orig_attr);

        return ret;
}
```

It became broken when the ret was changed - the inline wrappers really should rarely be changed.

So the logical step here is to make the op never NULL for new libibverbs and handle the fallback in verbs directly, as you are doing here.

But this wrapper doesn't test the input comp mask, while the cmd caller does, so it makes zero sense.

Let us go for worst case consistency then..

1) Fix ibv_query_device_ex inline wrapper to have

if (ret == EOPNOTSUPP || ret == ENOSYS)
    goto legacy;

This will allow new builds to work with old libibverbs again

2) Change the dummy query_device_ex to not fail and internally do to fallback. This will make old builds work with new libibverbs again

3) Make *both* the wrapper and the dummy match ibv_cmd_querty_device on the legacy path - check for input and comp_mask - we  can't have different flows with different requirements.

4) Continue to allow input to be NULL, I suppose this is ABI now.

5) Fix the man page for ibv_query_device_ex.3 which doesn't even have the right function signature or document really anything. Grr.

@Kamalheib your change stumbled into quite the mess here :(

Reply:
> @Kamalheib your change stumbled into quite the mess here :(

@jgunthorpe: Sorry to see that, well the change was suggested by you :/, please see the following:

https://github.com/linux-rdma/rdma-core/pull/659#discussion_r366458194

Reply:
> Sigh, this is a big mess it seems.
> 
I agree, 

> 
> Let us go for worst case consistency then..
> 
> 1. Fix ibv_query_device_ex inline wrapper to have
> 
> if (ret == EOPNOTSUPP || ret == ENOSYS)
> goto legacy;
> 
> This will allow new builds to work with old libibverbs again
you'll still have new builds return "EOPNOTSUPP" for ibv_query_device_ex and old libibverbs
will check for ENOSYS -> failing the operation.

> 
> 1. Change the dummy query_device_ex to not fail and internally do to fallback. This will make old builds work with new libibverbs again
do the fallback inside the query_device_ex ? 
that's what I did in the current pull-request.

> 2. Make _both_ the wrapper and the dummy match ibv_cmd_querty_device on the legacy path - check for input and comp_mask - we  can't have different flows with different requirements.
I don't follow here -> take the check outside of cmd_query? into the wrapper ? or leave it in the query_device as well ? 

> 3. Continue to allow input to be NULL, I suppose this is ABI now.
> 4. Fix the man page for ibv_query_device_ex.3 which doesn't even have the right function signature or document really anything. Grr.
> 
> @Kamalheib your change stumbled into quite the mess here :(

Reply:
#1 - The change in verbs.h must be as I said, ret == ENOSYS || ret == EOPNOTSUPP, can't delete that code.
#2 - Put it in all three places